### PR TITLE
Add notification permission for api level >=33 (fixes #1976)

### LINKED ISSUE #1976: Notifications not working in 1.25.0-rc.1.1
After updating to 1.25.0-rc.1.1, I'm no longer getting a notification from Syncthing. Looks like notifications are completely disabled for the application, as there is no configuration setting for notifications in Android App-Info for Syncthing (Settings -> Apps -> Syncthing). Other apps has a notification setting above Permissions in App-Info. This section is completely missing for Syncthing.

Could be a regression from #1972

### Version Information

    App Version: 1.25.0-rc.1.1
    Syncthing Version: v1.25.0-rc.1
    Android Version: 13

### PR DESCRIPTION
**Pull request by githubashutoshsoni**
Added notification permission.

### CODE REVIEW

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 1]
```diff
@@ -295,6 +298,16 @@ public Object instantiateItem(ViewGroup container, int position) {
             View view = layoutInflater.inflate(slides[position].layout, container, false);
 
             switch (slides[position]) {
+                case NOTIFICATION:
```
Comment (bt90) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:301]: You should also implement `shouldSkipSlide()` for `NOTIFICATION`

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 2]
```diff
@@ -389,6 +402,20 @@ private void requestLocationPermission() {
...
     }
 
+    private void requestNotificationPermission() {
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
+
+                requestPermissions(new String[]{Manifest.permission.POST_NOTIFICATIONS}, 1);
+
+            } else {
```
Comment (bt90) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:412]: I don't see a point in keeping an empty else block.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 3]
```diff
@@ -389,6 +402,20 @@ private void requestLocationPermission() {
                 Constants.PermissionRequestType.LOCATION.ordinal());
     }
 
+    private void requestNotificationPermission() {
```
Comment (bt90) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:405]:
```suggestion
    @TargetApi(33)
    private void requestNotificationPermission() {
```

File: app/src/main/res/layout/actiivty_notification.xml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (bt90) [inline app/src/main/res/layout/actiivty_notification.xml:1]: Could you mimic the the existing slides? e.g https://github.com/syncthing/syncthing-android/blob/main/app/src/main/res/layout/activity_firststart_slide_storage.xml

File: app/src/main/res/layout/actiivty_notification.xml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <TextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:padding="@dimen/desc_padding"
+        android:text="Notification will work if you enable it.  "
```
Comment (bt90) [inline app/src/main/res/layout/actiivty_notification.xml:11]: This should be moved into `strings.xml`

File: app/src/main/res/layout/actiivty_notification.xml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        android:textAppearance="?textAppearanceHeadline4"></TextView>
+
+    <Button
+        android:id="@+id/notification_btn"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="enabled notification"
```
Comment (bt90) [inline app/src/main/res/layout/actiivty_notification.xml:18]:
```suggestion
        android:text="@string/grant_permission"
```

**Review by bt90 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 1 above)
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:312]: fixed.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 2 above)
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: fixed

File: app/src/main/res/layout/actiivty_notification.xml (hunk 4 above)
Comment (githubashutoshsoni) [reply app/src/main/res/layout/actiivty_notification.xml:1]: fixed

File: app/src/main/res/layout/actiivty_notification.xml (hunk 6 above)
Comment (githubashutoshsoni) [reply app/src/main/res/layout/actiivty_notification.xml]: fixed

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 7]
```diff
@@ -191,6 +194,14 @@ private boolean isFirstStart() {
         return mPreferences.getBoolean(Constants.PREF_FIRST_START, true);
     }
 
+    private boolean isNotificationPermissionGranted() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```
Comment (bt90) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:198]:
```suggestion
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
        return true;
```

Early out. That way you don't need the permission check in the if block.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 8]
```diff
@@ -191,6 +194,14 @@ private boolean isFirstStart() {
         return mPreferences.getBoolean(Constants.PREF_FIRST_START, true);
     }
 
+    private boolean isNotificationPermissionGranted() {
```
Comment (bt90) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:197]: Maybe add the `@TargetAPI` ?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 9]
```diff
@@ -389,6 +413,19 @@ private void requestLocationPermission() {
                 Constants.PermissionRequestType.LOCATION.ordinal());
     }
 
+    @TargetApi(33)
+    private void requestNotificationPermission() {
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```
Comment (bt90) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:419]: Use the method above?

File: app/src/main/res/layout/actiivty_notification.xml (hunk 4 above)
Comment (bt90) [reply app/src/main/res/layout/actiivty_notification.xml:1]: Could you rename the file to match the others?

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -823,5 +823,8 @@ Please report any problems you encounter via Github.</string>
     <string name="browse">Browse</string>
     <string name="no_sub_folder_is_selected">No sub folder is selected</string>
     <string name="share_device_id_chooser">Share device ID with</string>
+    <string name="require_notification_permission_desc">Require notification permission to show notifications.</string>
```
Comment (bt90) [inline app/src/main/res/values/strings.xml:826]:
```suggestion
    <string name="require_notification_permission_desc">Syncthing requires your permission to show you notifictions.</string>
```

**Review by bt90 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 7 above)
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:198]: But we do need the else block to check if the permissions are granted or not?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 9 above)
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:421]: done

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 8 above)
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:198]: DOne

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 7 above)
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: But why? we do need else block for other checks anyhow? no?

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 11]
```diff
@@ -191,6 +194,16 @@ private boolean isFirstStart() {
         return mPreferences.getBoolean(Constants.PREF_FIRST_START, true);
     }
 
+    @TargetApi(33)
+    private boolean isNotificationPermissionGranted() {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
+            return true;
+        } else {
+            return ActivityCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED;
+        }
```
Comment (bt90) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:203]:
```suggestion
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            return true;
        }

        return ActivityCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED;
```

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 12]
```diff
@@ -389,6 +415,19 @@ private void requestLocationPermission() {
                 Constants.PermissionRequestType.LOCATION.ordinal());
     }
 
+    @TargetApi(33)
+    private void requestNotificationPermission() {
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```
Comment (bt90) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:421]: I think we can skip the check here. The method is only invoked if `isNotificationPermissionGranted()` returns `false`.

File: app/src/main/res/layout/actiivty_notification.xml (hunk 4 above)
Comment (bt90) [reply app/src/main/res/layout/actiivty_notification.xml:1]: `activity_firststart_slide_notification.xml` ?

**Review by bt90 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 11 above)
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:203]: Got it.

**Comment by bt90** [comment]
Thank you very much for working on this :)

I hope that I have enough time tonight to test the app changes.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 12 above)
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:421]: Got it.

**Comment by githubashutoshsoni** [comment]
Sure, @bt90 Let me know if anything else needs to be changed.

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -823,5 +823,8 @@ Please report any problems you encounter via Github.</string>
     <string name="browse">Browse</string>
     <string name="no_sub_folder_is_selected">No sub folder is selected</string>
     <string name="share_device_id_chooser">Share device ID with</string>
+    <string name="require_notification_permission_desc">Syncthing requires your permission to show you notifications.</string>
```
Comment (imsodin) [inline app/src/main/res/values/strings.xml:826]: This should describe why syncthing needs/wants to show the user notifications. Maybe something like this:

```suggestion
    <string name="require_notification_permission_desc">Syncthing requires your permission to show you notifications. Those are used to let you know about a new device connecting, or a new folder sharing request.</string>
```

File: app/src/main/res/drawable/baseline_notifications_active_24.xml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<vector android:height="24dp" android:tint="#000000"
+    android:viewportHeight="24" android:viewportWidth="24"
+    android:width="24dp" xmlns:android="http://schemas.android.com/apk/res/android">
+    <path android:fillColor="@android:color/white" android:pathData="M7.58,4.08L6.15,2.65C3.75,4.48 2.17,7.3 2.03,10.5h2c0.15,-2.65 1.51,-4.97 3.55,-6.42zM19.97,10.5h2c-0.15,-3.2 -1.73,-6.02 -4.12,-7.85l-1.42,1.43c2.02,1.45 3.39,3.77 3.54,6.42zM18,11c0,-3.07 -1.64,-5.64 -4.5,-6.32L13.5,4c0,-0.83 -0.67,-1.5 -1.5,-1.5s-1.5,0.67 -1.5,1.5v0.68C7.63,5.36 6,7.92 6,11v5l-2,2v1h16v-1l-2,-2v-5zM12,22c0.14,0 0.27,-0.01 0.4,-0.04 0.65,-0.14 1.18,-0.58 1.44,-1.18 0.1,-0.24 0.15,-0.5 0.15,-0.78h-4c0.01,1.1 0.9,2 2.01,2z"/>
```
[the remark this answers is not in the document]
Comment (githubashutoshsoni) [reply app/src/main/res/drawable/baseline_notifications_active_24.xml:4]: yes we are using SVG and it's pretty much scalable and looks pretty good. Can you try once running the app and letting me know if it works?

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (githubashutoshsoni) [reply app/src/main/res/values/strings.xml:826]: sure think looks more descriptive. let me add this.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 15]
```diff
@@ -53,7 +51,10 @@
 public class FirstStartActivity extends Activity {
 
     private enum Slide {
+
         INTRO(R.layout.activity_firststart_slide_intro),
+
+        NOTIFICATION(R.layout.activity_firststart_slide_notification),
         STORAGE(R.layout.activity_firststart_slide_storage),
         LOCATION(R.layout.activity_firststart_slide_location),
         API_LEVEL_30(R.layout.activity_firststart_slide_api_level_30);
```
[the remark this answers is not in the document]
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:60]: Done

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java [hunk 16]
```diff
@@ -63,7 +64,9 @@ private enum Slide {
         Slide(int layout) {
             this.layout = layout;
         }
-    };
+    }
+
+    ;
```
[the remark this answers is not in the document]
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:69]: must've been due to formatting.

File: app/src/main/res/drawable/baseline_notifications_active_24.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (bt90) [reply app/src/main/res/drawable/baseline_notifications_active_24.xml:4]:
```suggestion
    <path
        android:fillColor="?attr/colorControlNormal"
        android:pathData="M7.58,4.08L6.15,2.65C3.75,4.48 2.17,7.3 2.03,10.5h2c0.15,-2.65 1.51,-4.97 3.55,-6.42zM19.97,10.5h2c-0.15,-3.2 -1.73,-6.02 -4.12,-7.85l-1.42,1.43c2.02,1.45 3.39,3.77 3.54,6.42zM18,11c0,-3.07 -1.64,-5.64 -4.5,-6.32L13.5,4c0,-0.83 -0.67,-1.5 -1.5,-1.5s-1.5,0.67 -1.5,1.5v0.68C7.63,5.36 6,7.92 6,11v5l-2,2v1h16v-1l-2,-2v-5zM12,22c0.14,0 0.27,-0.01 0.4,-0.04 0.65,-0.14 1.18,-0.58 1.44,-1.18 0.1,-0.24 0.15,-0.5 0.15,-0.78h-4c0.01,1.1 0.9,2 2.01,2z"/>
```

[the remark this answers is not in the document]
Comment (bt90) [reply app/src/main/res/drawable/baseline_notifications_active_24.xml:4]: See https://medium.com/androiddevelopers/android-styling-common-theme-attributes-8f7c50c9eaba

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (imsodin) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:69]: Yeah but the new formatting can't be it, it's pretty ridiculous, isn't it? Can you revert it please.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (imsodin) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Other places aren't adjust yet, e.g. here (not limited to that place): https://github.com/syncthing/syncthing-android/pull/1983/files#diff-7ad918fdbb730c99ff756cc4a47f080cd25118fbac4768655532d8d8b7728525R243

File: app/src/main/res/drawable/baseline_notifications_active_24.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (githubashutoshsoni) [reply app/src/main/res/drawable/baseline_notifications_active_24.xml:4]: Got it. Thanks

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java:69]: Okay

File: app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: Got it. I think I've completed this now. let me know any code snippets I've missed.

[the remark this answers is not in the document]
Comment (githubashutoshsoni) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java]: I eagerly anticipate accepting the pull request, considering the occasion is Hacktoberfest and I have three additional pull requests to complete. Thank you!

**Review by imsodin — APPROVED** [review]
lgtm

Waiting for @bt90's feedback before merging.

**Comment by githubashutoshsoni** [comment]
Thanks @imsodin. @bt90 let me know if this is good enough for the pull request?

**Review by bt90 — APPROVED** [review]
LGTM.

Screenshot_20231011-070459~2

**Comment by githubashutoshsoni** [comment]
> lgtm
> 
> Waiting for @bt90's feedback before merging.

Thanks @bt90 for approving. Can this be merged now? @imsodin

**Pull request merged**